add a multiselect 2.0 spec (#227)
libp2p/specs: connections/multiselect2.md

# Multiselect 2.0

## Introduction

Multiselect 2.0 replaces the Multistream protocol. Compared to its predecessor, it offers:

1. Downgrade protection for the security protocol negotiation.
2. Zero-rountrip stream multiplexer negotiation for handshake protocols that take advantage of early data mechanisms (one-roundtrip negotation for protocols / implementations that don't).
3. Compression for the protocol identifiers of frequently used protocols.

By using protobufs for all control messages, Multiselect 2.0 provides an easy path for future protocol upgrades. The protobuf format guarantees that unknown field in a message will be skipped, thus future version of the protocol can add new fields that signal support for new protocol features.

## High-Level Overview

> ### Handshake Protocol Selection

Reviewer: Let's call this Secure Channel Selection, to keep it consistent with the rest of the spec corpus.

> Handshake protocols are not being negotiated, but are announced in the peers' multiaddrs. 

Author: What is that corner case you're referring to? I'm not sure if we should even build that flexibility into our stack, even if we could come up with a situation where a negotiation wouldn't be susceptible to a MITM attack. If it's not there, it can't be exploited.

> Handshake protocols are not being negotiated, but are announced in the peers' multiaddrs. 

Reviewer: Proposal for prose:

> Conversely to multistream-select 1.0, secure channel protocols are not dynamically negotiated in multiselect 2.0. Instead, they are announced upfront in the peer multiaddrs. It is up to the implementers to decide whether each secure channel is exposed over a different port, or if a single port handles all secure channels, and a trial-and-error strategy is used to deserialise and identify which protocol is being used.

Author: Applied with minor rewording.

> **TODO**: Do we need to describe the format here? I guess we don't, but we will probably need another document for that change, and we can link to it from here.

Reviewer: We need an extension to the multiaddr spec + multicodec table to add secure channels as an atom. Also, @yusefnapora was writing this document to specify the semantics of multiaddrs: https://github.com/libp2p/specs/pull/191.

Peers advertising a multiaddr that includes a handshake protocol MUST support Multiselect 2.0 as described in this document.

#### TCP Simultaneous Open

TCP allows the establishment of a connection if two endpoints start initiating a connection at the same time. This is called TCP Simultaneous Open. For libp2p, this is problematic, since most stream multiplexers assign stream IDs based on the role (client or server) of and endpoint.

> It is therefore desirable to fail a connection attempt as early as possible, if a TCP Simultaneous Open occurs. TLS 1.3 and Noise provide this guarantee: For example, the TLS handshake will fail if an endpoint receives a ClientHello instead of a ServerHello as a response to its ClientHello.

Reviewer 2: @vyzo, is it true that we want to fail the connection completely if a TCP simultaneous open occurs? I remember you were thinking though how to do NAT hole-punching with TCP and that simultaneous open was part of that story. 

It seems like for hole punching, we want one of the simultaneous connection attempts to succeed, but we need a deterministic way to figure out which party has the initiator role. I don't know how that fits in to this multiselect design, since multiselect now only kicks in _after_  the crypto handshake, and it seems like TLS and Noise would both just fail the handshake if both parties believe they are the initiator.

Reviewer: One of the requirements of ms2.0 is to be deep-packet inspection censorship resistant. This means not exchanging easily recognisable traffic in plaintext (e.g. secure channel negotiation, or tokens like `iamclient`).

We now advertise secure channels in the multiaddr, and the initiator jumps right into the handshake.

IMO we need a conflict resolution mechanism that sits above multiselect 2.0, detects TCP simultaneous connects, and resolves the conflict via some deterministic or probabilistic heuristic.

An example of a deterministic heuristic could be:

> If we are the initiator and we receive an invalid payload from the responder (plus points if we can detect it was a ClientHello – TLS 1.3 – or the wrong Noise message), both sides hash the outgoing bytes and incoming bytes exchanged until then, and the highest value becomes the initiator. The handshake is tried again with each side taking their new role.

That scheme would require us to "tee" the `io.Reader` and `io.Writer` until the handshake is over, which is probably fine.

An example of a probabilistic heuristic could be:

> If we are the initiator and receive an invalid payload from the responder, we roll a dice to determine if we're initiator or responder, and try again, up to a maximum of 3 times.

Since secio doesn't provide this property, secio cannot be used with Multiselect 2.0.

### Stream Multiplexer Selection

This section only applies if Multiselect 2 is run over a transport that is not natively multipexed. Transports that provide stream multiplexing on the transport layer (e.g. QUIC) don't need to do anything described in this section.

> Some handshake protocols (TLS 1.3, some variants of Noise (**TODO**: specify which)) support sending of *Early Data*. Early Data can be sent by the server after receiving the first handshake message from the client. It is encrypted, however, at that point of the handshake the client's identity is not yet verified.

Reviewer: Clarification: all Noise messages can carry application payloads. The security guarantees of the payload vary depending on the stage of the handshake we send it in.

> Some handshake protocols (TLS 1.3, some variants of Noise (**TODO**: specify which)) support sending of *Early Data*. Early Data can be sent by the server after receiving the first handshake message from the client. It is encrypted, however, at that point of the handshake the client's identity is not yet verified.

Author: Thanks. I updated the text about Noise, prohibiting the use of unencrypted data for announcing the stream multiplexers, but allowing it as soon as data can be encrypted *somehow* (even if the peer's identity is not verified at that moment). This should give us roughly the same security guarantees as we're getting with TLS 1.3.

In Multiselect 2 the server makes use of Early Data by sending a list of stream multiplexers. This ensures that the client can choose a stream multiplexer as soon as the handshake completes (or fail the connection if it doesn't support any stream multiplexer offered by the server).

Note that this negotiation scheme allows peers to negotiate a "monoplexed" connection, i.e. a connection that doesn't use any stream multiplexer. Endpoints can offer support for monoplexed connections by offering the `/monoplex` stream multiplexer.

**TODO**: Do we need to define a way to send an error code / error string? Or do we have something like that in libp2p already?

![](handshake.png)

Handshake protocols (or implementations of handshake protocols) that don't support sending of Early Data will have to run the stream multiplexer selection after the handshake completes.

#### 0-RTT

> When using 0-RTT session resumption as offered by TLS 1.3 and some variants of Noise (**TODO**: specify which), the endpoints MUST remember the negotiated stream multiplexer used on the original connection. This ensures that the client can send application data in the first flight when resuming a connection.

Reviewer: Take into account that multiplexers can change across restarts.

> When using 0-RTT session resumption as offered by TLS 1.3 and some variants of Noise (**TODO**: specify which), the endpoints MUST remember the negotiated stream multiplexer used on the original connection. This ensures that the client can send application data in the first flight when resuming a connection.

Author: @raulk That's on us to define. We could say that if you resume a connection, you have to use the same multiplexer. This would save us a roundtrip for negotiation the stream multiplexer in the common case.
In the rare case when a peer dropped support for a muxer between two connections, you would just reject 0-RTT and fall back to the 1-RTT handshake.

TLS as well as Noise will fail the handshake if both endpoints act as clients. In case of such a handshake failure, the two endpoints need to restart the handshake. Endpoints MUST NOT close the underlying TCP connection in this case. Implementations SHOULD specifically test for this type of handshake failure, and not treat any handshake failure as a potential Simultaneous Open.

To determine the roles in the second handshake attempt, endpoints calculate the SHA-256 hash of the handshake messages that were sent and received (including any error message(s) that the handshake protocol might have sent) during the failed handshake attempt.
The peer that sent the messages resulting in the numerically smaller hash value acts a client in the second handshake attempt, the peer that sent the messages resulting in the numerically larger hash value acts as a server.

> Since secio assign roles during the handshake, it is not possible to detect a Simultaneous Open in this case. Therefore, secio MUST NOT be used with Multiselect 2.0.

Reviewer 3: ok, this is not quite right; we need something for secio too, unless we are actively deprecating it.

Author: Which part of it is not right?
We've been talking about deprecating it for a long time, and the main reason we've been sticking to it is for backwards compatibility. Since Multiselect 2.0 is not backwards compatible anyways, this would be a good opportunity to finally phase it out.

Reviewer 4: I agree with the push towards deprecation, as seems to be the growing consensus. Multiselect 2.0 presents a convenient opportunity to upgrade our secure channels.

Reviewer 5: There is no path for backwards compatibility here with multistream 1 correct? Pairing the secio deprecation with multiselect 2 is going to put a lot of stress on limited environments (browsers) to upgrade, or segregate them from the rest of the network. I don't think it warrants blocking the spec, especially if providing that compatibility hampers the performance/feature gains, but we should be cognizant and clear of the network rollout time table and its impact on the network.

Reviewer 5: secio is actively being deprecated and removed from the network, so this no longer needs to be a consideration.

Reviewer 6: @jacobheun @raulk 

IIUC, even Noise needs an initiator and a responder. So, even after deprecating SecIO, we still need to assign roles.

Author: I'm happy to update the document and remove any mention of secio, provided there's interest in moving forward with this document. It's been quiet for long time...

# Multiselect 2.0

## Introduction

Multiselect 2.0 replaces the Multistream protocol. Compared to its predecessor, it offers:

1. Downgrade protection for the security protocol negotiation.
2. Zero-rountrip stream multiplexer negotiation for handshake protocols that take advantage of early data mechanisms (one-roundtrip negotation for protocols / implementations that don't).
3. Compression for the protocol identifiers of frequently used protocols.

By using protobufs for all control messages, Multiselect 2.0 provides an easy path for future protocol upgrades. The protobuf format guarantees that unknown field in a message will be skipped, thus future version of the protocol can add new fields that signal support for new protocol features.

## High-Level Overview

### Secure Channel Selection

> Conversely to multistream-select 1.0, secure channel protocols are not dynamically negotiated in multiselect 2.0. Instead, they are announced upfront in the peer multiaddrs. It is up to the implementers to decide whether each secure channel is exposed over a different port, or if a single port handles all secure channels, and a demultiplexing strategy is used to identify which protocol is being used.

Reviewer: ```suggestion
Conversely to multistream-select 1.0, secure channel protocols are not dynamically negotiated in-band. Instead, they are announced upfront in the peer multiaddrs (<add link to multiaddr spec>). This way, implementations can jump straight into a cryptographic handshake, thus curtailing the possibility of packet-inspection-based censorship and dynamic downgrade attacks.

It is up to the implementers to decide whether each secure channel is exposed over a different port, or if a single port handles all secure channels, and a demultiplexing strategy is used to identify which protocol is being used.
```

**TODO**: Do we need to describe the format here? I guess we don't, but we will probably need another document for that change, and we can link to it from here.

> Peers advertising a multiaddr that includes a handshake protocol MUST support Multiselect 2.0 as described in this document.

Reviewer: I know we had discussed this assumption as a way to simplify things, but I do think it'll end up being short-sighted, and in some ways, a regression compared to multistream-select 1.0, which does announce its version (although admittedly the implementations are not ready to support multiple versions, but the protocol is).

Something I've been thinking about is to create an extension to multistream-select 1.0 / upgraders that would allow us to go straight into a cryptographic handshake, as a way to deliver censorship resistance to downstream users that require it before we realistically ship ms2.0.

Author: I'm not sure how this would work without either defining another multicoded or requiring an additional round-trip to negotiate. I'd prefer to keep ms2.0 as simple as possible, which will allow us to ship it faster.

> #### TCP Simultaneous Open

Reviewer: From the perspective of this spec, we should not talk about specific transports, but rather about "role deduplication" or something like that. The TCP simultaneous open is a specific instance where this deduplication is necessary, and at the connection bootstrapping layer, the transport mechanics are irrelevant.

Author: @tomaka already raised that point in https://github.com/libp2p/specs/pull/227#discussion_r350673105, and we agreed to move the actual logic to the transport / secure channel specification, while requiring these to define *some way* to deal with TCP sim open in this document.

TCP allows the establishment of a connection if two endpoints start initiating a connection at the same time. This is called TCP Simultaneous Open. For libp2p, on the one hand, this poses a problem, since most stream multiplexers assign stream IDs based on the role (client or server) of an endpoint. On the other hand, it can be used as a hole punching technique to facilitate NAT traversal.

TLS as well as Noise will fail the handshake if both endpoints act as clients. In case of such a handshake failure, the two endpoints need to restart the handshake. Endpoints MUST NOT close the underlying TCP connection in this case. Implementations SHOULD specifically test for this type of handshake failure, and not treat any handshake failure as a potential Simultaneous Open.

> To determine the roles in the second handshake attempt, endpoints compare the SHA-256 hashes of their peer IDs. The peer with the numerically smaller hash value acts as a client in the second handshake attempt, the peer with the numerically larger hash value acts as a server.

Reviewer 7: This forces us to transfer our peer ID in the initial message, right?
While this is normally good, it also means we can't for example fool proxies by pretending that our traffic in regular HTTP3.

Reviewer 7: Handling TCP simultaneous connections seems a bit off-topic for multistream-select, and could be specific to each transport and/or encryption protocol.

Author: When you dial a peer, you already know it’s peer ID. And of course you also know your own peer ID, so we don’t have to send anything extra on the wire.

Author: @tomaka I tend to agree with this being slightly off-topic. The reason it's in here is because ms-2.0 also changes the connection bootstrapping, and #196 won't work any more (at least not if we want to be able to mask our traffic).

What do you think of saying in this document 
> Secure Channels MUST define how to handle TCP simultaneous open, if they can be used over TCP.

and then moving the text here to the TLS document?

Reviewer 7: > When you dial a peer, you already know it’s peer ID. And of course you also know your own peer ID, so we don’t have to send anything extra on the wire.

The receiving side doesn't know the dialer's peer ID, though!

While maybe not very important, there's also the use-case where you just want to connect to a node you know is at a given IP address without bothering fetching its peer ID.

Reviewer: @marten-seemann 

> When you dial a peer, you already know it’s peer ID. And of course you also know your own peer ID, so we don’t have to send anything extra on the wire.

I may be missing something, but how does the responder learn the peer ID without a hard requirement for the crypto handshake to transmit it on the first message?

Reviewer: I think there are two different specs here:

* one for connection bootstrapping: crypto channel, multiplexer intersection, connection-scoped feature flags, role-deduplication (TCP simultaneous open is just one instance of a case where such deduplication is needed).
* one for stream-wise protocol negotiation: proper multiselect 2.0 in my head.

The hypothesis is that having a holistic discussion is easier, because this spec is the spiritual successor of multistream-select 1.0 which participates in both, but materially land it as two different specs.

Reviewer 3: It's a simultaneous dial, you have both peer IDs.

Reviewer: At this layer, you only know you have established a connection and have exchanged some bytes. You *think* you know the other side’s peer ID, but it’s not authenticated. After exchanging those initial bytes, you notice the other party sends you a message you didn’t expect. Does each party work on their assumption of the peer ID of the other party? That leaves the system open to a series of attacks where a peer advertises a very large/small peer ID for itself, then responds to all SYN packets with another SYN + the initiator message, to force a conflict resolution pathway it knows it’ll always win. I much rather introduce a vector of randomness per-session.

Author: @raulk What's the attack here? Sure, an attacker can go through the hassle you're describing to make sure it ends up in the client role (from the viewpoint of the cryptographic handshake). During the cryptographic handshake both peer validate each other's peer IDs, so it seems to me that the attacker gains nothing from this attack at all.

Reviewer: I'm not worried about peers _lying_ about their identity. I'm worried about making the conflict resolution 100% deterministic, and the transitive attack surface that might expose going forward. We're specifying against our future selves, not against the situations we foresee now.

If there's a bug exploitable in this circumstance, an attack can be crafted that works 100% of the time, by precomputing a large/small peer identity.

I admit I may be thinking too far. But if we want to make this conflict resolution probabilistic (which I think is the right way from first principle), when we can make the protocol can send a random nonce and have peers XOR their nonces to calculate who wins, or something like that.

Author: I’m not sure I understand the argument here. The security of our handshake relies on the TLS handshake. If TLS is broken, we don’t gain a lot from reducing the cost of an attack from succeeding in 100% of the cases to 50% of the cases.

Since secio assign roles during the handshake, it is not possible to detect a Simultaneous Open in this case. Therefore, secio MUST NOT be used with Multiselect 2.0.

> ### Stream Multiplexer Selection

Reviewer: I would make this section about the pattern we're adopting:

1. use early message data.
2. share multiplexers in order of precedence, in the first handshake message that supports encrypted early data.
3. intersect the result.

In a separate section, we can norm on the concrete implementation with TLS 1.3 and Noise (and those specs needs to be enhanced accordingly).

Question: can we assume that all handshakes now, and in the future, will support one encrypted early data per peer (A -> B, B -> A)? I don't think we can count on that, so we need a safe fallback.

Author: I think this is pretty much guaranteed from the way a key exchange works, that the server will end up with the keys half a round-trip before the client.

If we want to do the intersection of stream multiplexers you're suggesting, we'll need to specify a new protobuf message that's *only* used for that purpose. I'm not opposed to it, but I remember that @Stebalien had a strong preference for using the same mechanism for all protocol selections (i.e. not making a difference between muxer and application protocol negotiation).

Author: See #230.

> This section only applies if Multiselect 2 is run over a transport that is not natively multipexed. Transports that provide stream multiplexing on the transport layer (e.g. QUIC) don't need to do anything described in this section.

Reviewer 5: ```suggestion
This section only applies if Multiselect 2 is run over a transport that is not natively multiplexed. Transports that provide stream multiplexing on the transport layer (e.g. QUIC) don't need to do anything described in this section.
```

Some handshake protocols (TLS 1.3, Noise) support sending of *Early Data*. 

> In Multiselect 2 the server makes use of Early Data by sending a list of stream multiplexers. This ensures that the client can choose a stream multiplexer as soon as the handshake completes (or fail the connection if it doesn't support any stream multiplexer offered by the server).

Reviewer: I'd word similar as follows:

> Both parties share the list of supported stream multiplexers in the earliest handshake message that supports encrypted early data, listed in order of precedence. Both parties intersect the list and choose the highest-ranking shared multiplexer, or abort the connection by sending a `DISCONNECT` message (we need to specify it) if no mutually supported multiplexer exists.

Author: That's a new proposal how stream multiplexer negotiation should work. The spec doesn't define any intersecting of lists, and therefore doesn't need to specify an order of preference.

Author: See #230.

Author: Resolved by merging #230.

When using TLS 1.3, the server can send Early Data after it receives the ClientHello. Early Data is encrypted, but at this point of the handshake the client's identity is not yet verified.

> While Noise in principle allows sending of unencrypted data, endpoints MUST NOT use this to send their list of stream multiplexers. An endpoint MAY send it as soon it is possible to send encrypted data, even if the peers' identity is not verified at that point.

Reviewer: A `MAY` here suggests that multiplexers may never get sent in Noise. How do we handle that case?

Author: It's MAY send as soon ..., otherwise it will send it later.

> While Noise in principle allows sending of unencrypted data, endpoints MUST NOT use this to send their list of stream multiplexers. An endpoint MAY send it as soon it is possible to send encrypted data, even if the peers' identity is not verified at that point.

Reviewer 5: ```suggestion
While Noise in principle allows sending of unencrypted data, endpoints MUST NOT use this to send their list of stream multiplexers. An endpoint MAY send it as soon as it is possible to send encrypted data, even if the peers' identity is not verified at that point.
```

Note that this negotiation scheme allows peers to negotiate a "monoplexed" connection, i.e. a connection that doesn't use any stream multiplexer. Endpoints can offer support for monoplexed connections by offering the `/monoplex` stream multiplexer.

**TODO**: Do we need to define a way to send an error code / error string? Or do we have something like that in libp2p already?

![](handshake.png)

> Handshake protocols (or implementations of handshake protocols) that don't support sending of Early Data will have to run the stream multiplexer selection after the handshake completes.

Reviewer: Oh ok, I see that you added the fallback. We need to define how that'll work.

Author: I'm not sure if there's anything we need to define. "Early data" is not a separate byte stream, it's the same byte stream as the rest of the connection. The only difference is that the data is sent earlier (and, depending on the handshake protocol, might use a different set of keys).
To the application Early Data and Late (?) Data is not distinguishable at all.

#### 0-RTT

> When using 0-RTT session resumption as offered by TLS 1.3 and Noise, the endpoints MUST remember the negotiated stream multiplexer used on the original connection. This ensures that the client can send application data in the first flight when resuming a connection.

Reviewer: 1. Stream multiplexers can change over time.
2. Peers may forget other peers' protocols.

We may need some form of opaque token with which we can verify that our assertions about the other party still remain valid, upon reconnecting. If the other party NACKs, we fall back to full connection bootstrapping.

Author: The TLS session ticket is exactly that opaque token.

Reviewer: But multiselect 2.0 doesn't know about concrete secure channels. If there's a handshake-specific token that we can bind to, it needs to percolate up to this layer.

> ## Protocol Speficiation

Reviewer: ```suggestion
## Protocol Specification
```

All messages are Protobuf messages using the `proto3` syntax. Every message is wrapped by the `Multiselect` message:

```protobuf
# Wraps every message
message Multiselect {
    oneof message {
        Offer offer = 1;
        Use use = 2;
    }
}

> ```

Reviewer 7: Nit: for readability I'd put the definition of `Offer` and `Use` here and not below.

Reviewer: Yes, we need the definitions and protocol overview before the wire format.

Reviewer: And can we have the entire schema dumped at once after the overview. It's difficult to reason when types definitions are split across prose.

Author: Agreed. I'll do another editorial pass over the document soon, and will apply this suggestion.

Author: I did another editorial pass over the document and moved the protobuf definitions to a separate section.

The `Offer` message is used to initiate a conversation on a new stream. It contains either a single or multiple protocols that the endpoint would like to use on the stream.
A `Protocol` is the application protocol spoken on top of an ordered byte stream. The `name` of a protocol is the protocol identifier, e.g. `/ipfs/ping/1.0.0`. The `id` is a numeric abbreviation for this protocol (see below for details how `id`s are assigned).
If the endpoint only selects a single protocol, it MAY start sending application data right after the protobuf message. Since it has not received confirmation if the peer actually supports the protocol, any such data might be lost in that case.
If the endpoint selects multiple protocols, it MUST wait for the peer's choice of the application protocol (see description of the `Use` message) before sending application.

```protobuf
# Select a list of protocols.
message Offer {
    message Protocol {
        oneof protocol {
            string name = 1;
            uint64 id = 2;
        }
    }
    repeated Protocol protocols = 1;
}
```

> The `Use` message is sent in response to the `Offer`. And endpoint MUST treat the receipt of a `Use` message before having sent a `Offer` message on the stream as a connection error.

Reviewer 5: ```suggestion
The `Use` message is sent in response to the `Offer`. An endpoint MUST treat the receipt of a `Use` message before having sent an `Offer` message on the stream as a connection error.
```

> If none of the protocol(s) listed in the `Offer` message are acceptable, and endpoint resets both the send- and the receive-side of the stream.

Reviewer 5: Should the reset be required?
```suggestion
If none of the protocol(s) listed in the `Offer` message are acceptable, an endpoint MUST reset both the send- and the receive-side of the stream.
```

> If an endpoint receives an  `Offer` message that only offers a single protocol, it accepts this protocol by sending an empty `Use` message (i.e. a message that doesn't list any `protocol`), or a `Use` message that assigns a protocol id (see below). Sending an empty `Use` message in response to a`Offer` message that offers multiple protocols is not permitted, and MUST be treated as a connection error by an endpoint.

Reviewer 5: ```suggestion
If an endpoint receives an  `Offer` message that only offers a single protocol, it accepts this protocol by sending an empty `Use` message (i.e. a message that doesn't list any `protocol`), or a `Use` message that assigns a protocol id (see below). Sending an empty `Use` message in response to an `Offer` message that offers multiple protocols is not permitted, and MUST be treated as a connection error by an endpoint.
```

> If an endpoint receives a `Offer` message that offers multiple protocols, it chooses an application protocol that it would like to speak on this stream. It informs the peer about its choice by sending its selection in the `protocol` field of the `Use` message.

Reviewer 5: ```suggestion
If an endpoint receives an `Offer` message that offers multiple protocols, it chooses an application protocol that it would like to speak on this stream. It informs the peer about its choice by sending its selection in the `protocol` field of the `Use` message.
```

#### TCP Simultaneous Open

TCP allows the establishment of a connection if two endpoints start initiating a connection at the same time. This is called TCP Simultaneous Open. For libp2p, on the one hand, this poses a problem, since most stream multiplexers assign stream IDs based on the role (client or server) of an endpoint. On the other hand, it can be used as a hole punching technique to facilitate NAT traversal.

TLS as well as Noise will fail the handshake if both endpoints act as clients. In case of such a handshake failure, the two endpoints need to restart the handshake. Endpoints MUST NOT close the underlying TCP connection in this case. Implementations SHOULD specifically test for this type of handshake failure, and not treat any handshake failure as a potential Simultaneous Open.

> To determine the roles in the second handshake attempt, endpoints calculate the SHA-256 hash of the handshake messages that were sent and received (including any error message(s) that the handshake protocol might have sent) during the failed handshake attempt.

Reviewer 3: Are these handshake messages available to the application? It seems to me that these will be internal to the TLS implementation and not actually available to hash.

Author: You can just wrap the `Read` and `Write` of the `net.Conn`.

Reviewer 3: This is true in go, but what about other languages?
Also, does the "wrap the Read/Write" imply that we hash all the bytes sent/received to make the decision?
It's still now clear how an implementor would implement this part of the spec.

Author: These are the bytes sent in the clear over the wire. I'm not familiar with the TLS / Noise implementations in other languages, but I'd be surprised if this posed a major challenge in any language.
@tomaka, @yusefnapora Can you shed some light on this?

Reviewer 3: Any language that uses a TLS library that reads/writes to the file descriptor will have problems with this.

Reviewer 3: We discussed this with @Stebalien and we both think this (hashing the plaintext messages) is the wrong way to approach it.
We totally don't have to do that: we have the peer IDs for both endpoints here, because we need them for the TLS handshake. 
So what we should do is simply pick the peer with the highest (or lowest) numeric peer id to act as the initiator if the handshake fails with unexpected client hello.

Author: @vyzo Thank you. Using peer IDs makes a lot of sense to me, and is definitely much simpler than recording messages sent on the wire.
I updated the spec using SHA256(peer ID). I believe that makes numerical comparisons of peer IDs easier, especially if they have different lengths.

Author: There's an additional detail we need to specify: When receiving the unexpected handshake message, TLS does two things:
1. It sends a TLS alert to the peer, and
2. it aborts the handshake and returns an error message.

If we just restart the handshake when the handshake message is returned, the peer's alert will still be in flight, and typically will be received after we initiated the second handshake, leading to a handshake failure. There are three ways we can solve this:
1. Modify the TLS stack to not send the alert,
2. Filter out the alert before it reaches the TLS stack
3. Wait for the alert to arrive, and then start the second connection attempt afterwards.

I don't think 1. is very practical, since it requires us to modify the TLS stack. 2. would be doable, but requires wrapping of the connection. 3. costs us half a roundtrip, but could be implemented without modifications to the TLS stack or wrapping of the connection.

Author: Update: I just implemented option 3 in https://github.com/libp2p/go-libp2p-tls/compare/tcp-sim-open-option3, and unfortunately, it's inherently racy: When `tls.Conn().Handshake()` returns, we have no idea if this is before or after the TLS alert from the peer was received. We therefore can't reliably catch the TLS alert afterwards, since it might never arrive.

I think there's an option 4:
4. Filter out the `ClientHello` before it arrives at the TLS stack. If we receive one, abort the current handshake (taking care that no abort alert is sent out), and start anew.

Unfortunately, this option requires wrapping of the TLS connection, but at least it should be race-free. Furthermore, it doesn't cost us any additional (half) round-trips.

Author: Here's an implementation of option 4, which seems to work: https://github.com/libp2p/go-libp2p-tls/pull/38.
The downside is that we're duplicating a bit of the record layer parsing code of TLS, since we need to make sure that we read the whole `ClientHello` from the connection before we can start the second connection attempt.

Reviewer 3: The concern here is again other language implementations and whether they can do that in reasonable fashion.
Other than that, I am fine with it.

Reviewer 2: I haven't verified this at all, so we shouldn't assume this is correct... but the [docs for node's tls.connect()](https://nodejs.org/api/tls.html#tls_tls_connect_options_callback) show that you can pass in any duplex Stream as the underlying socket which will be used for the connection. 

So it seems like we should be able to wrap a `net.Socket` with some code similar to @marten-seemann's go code which parses the first handshake message and aborts with a custom error if the server sends a `ClientHello`. Then we can do the peer id comparison and retry the connection when we get the special error type.

This won't work in browsers, but I don't think we have a clear path to supporting TLS 1.3 at all there, and we don't have access to raw TCP in browsers anyway, so it's kind of a moot point.
